IPAddress.GetHashCode based on Marvin instead of ToString

i3arnon opened the pull request:
This addresses #24117.

I've used [`System.Numerics.Hashing.HashHelpers`](https://github.com/dotnet/corefx/blob/master/src/Common/src/System/Numerics/Hashing/HashHelpers.cs) instead of writing a custom implementation as it's deemed good enough for the various `ValueTuple`s.

I've used BenchmarkDotNet to compare the current and suggest implementations without caching the results on `GetHashCode` or `ToString` as that represents the more interesting scenario: Invoking `GetHashCode` on a constant stream of new `IPAddress` instances. The differences become less relevant the more instances are being reused.

The code for the comparison is [here](https://gist.github.com/i3arnon/e9f85402790fcbd1d14450540f7ec496) and the results are:

``` ini

BenchmarkDotNet=v0.10.9, OS=Windows 10 Redstone 2 (10.0.15063)
Processor=Intel Core i7-7600U CPU 2.80GHz (Kaby Lake), ProcessorCount=4
Frequency=2835941 Hz, Resolution=352.6166 ns, Timer=TSC
.NET Core SDK=2.0.0
  [Host]     : .NET Core 2.0.0 (Framework 4.6.00001.0), 64bit RyuJIT
  DefaultJob : .NET Core 2.0.0 (Framework 4.6.00001.0), 64bit RyuJIT


```
 |                 Method |       Mean |     Error |    StdDev |  Gen 0 | Allocated |
 |----------------------- |-----------:|----------:|----------:|-------:|----------:|
 |    GetHashCodeToString | 182.927 ns | 3.5535 ns | 3.4900 ns | 0.0341 |      72 B |
 | GetHashCodeHashHelpers |   6.685 ns | 0.1681 ns | 0.1936 ns |      - |       0 B |

dnfclas commented:
__@i3arnon__, thanks for signing the contribution license agreement. We will now validate the agreement and then the pull request.
<br />_Thanks, .NET Foundation Pull Request Bot_

davidsh commented:
cc: @karelz

karelz commented on src/System.Net.Primitives/src/System/Net/IPAddress.cs:
```diff
@@ -614,7 +615,13 @@ public override int GetHashCode()
+                    var hashCode = HashHelpers.Combine(HashHelpers.RandomSeed, unchecked((int)PrivateScopeId));
+                    for (int i = 0; i < NumberOfLabels; i += 2)
+                    {
+                        hashCode = HashHelpers.Combine(hashCode, _numbers[i] << sizeof(ushort) | _numbers[i + 1]);
```
sizeof(ushort) is in bytes, not bits -- 8*sizeof(ushort)

i3arnon replied:
Of course it is :).
I had it at 16 originally. That's what I get for trying to be smart.

karelz commented on src/System.Net.Primitives/src/System/Net/IPAddress.cs:
```diff
@@ -614,7 +615,13 @@ public override int GetHashCode()
                 if (_hashCode == 0)
                 {
-                    _hashCode = StringComparer.OrdinalIgnoreCase.GetHashCode(ToString());
+                    var hashCode = HashHelpers.Combine(HashHelpers.RandomSeed, unchecked((int)PrivateScopeId));
```
Nit: Replace vat with int. See our coding style - if the type is not clear from the line, we should not use var.

i3arnon replied:
Got it. Done

justinvp commented:
The JIT will likely generate better code if you cache `_numbers` in a local and check the array length, e.g.:

```c#
Debug.Assert(_numbers != null && _numbers.Length == NumberOfLabels);
int hashCode = HashHelpers.Combine(HashHelpers.RandomSeed, unchecked((int)PrivateScopeId));
ushort[] numbers = _numbers;
for (int i = 0; i < numbers.Length; i += 2)
{
    hashCode = HashHelpers.Combine(hashCode, numbers[i] << 16 | numbers[i + 1]);
}
```

danmosemsft commented on src/System.Net.Primitives/src/System/Net/IPAddress.cs:
```diff
@@ -614,7 +615,13 @@ public override int GetHashCode()
             {
                 if (_hashCode == 0)
                 {
```
comment above still refers to generating hash from string repr'n

i3arnon replied:
Thanks, I updated it to match the new implementation.

danmosemsft commented on src/System.Net.Primitives/src/System/Net/IPAddress.cs:
```diff
@@ -614,7 +615,13 @@ public override int GetHashCode()
                 if (_hashCode == 0)
                 {
-                    _hashCode = StringComparer.OrdinalIgnoreCase.GetHashCode(ToString());
+                    int hashCode = HashHelpers.Combine(HashHelpers.RandomSeed, unchecked((int)PrivateScopeId));
```
I assume the random seed is to try to make hash codes less predictable to avoid DoS attacks. Fine. I do notice that in only one other place we use HashHelpers.Combine do we use the random seed.

danmosemsft replied:
If we care about DoS attacks then we should do the same on the ipv4 calculation below.

@morganbr how much do we want to use randomized hashing on arbitrary types? Or basically just strings?

i3arnon replied:
That's the idea, yes. AFAIK currently there are only 2 usages of HashHelper, one with a random seed and the other without.

danmosemsft replied:
There are several others https://github.com/dotnet/corefx/search?utf8=%E2%9C%93&q=hashhelpers.combine&type=

i3arnon replied:
I mainly looked at `ValueTuple`...
Should I remove the seed?

danmosemsft replied:
What do folks put these into hashtables for? Is it likely to be something that untrusted clients have influence over? The usual DoS concern is people supplying specially crafted strings.

Either way, we should do the same for ipv4.

danmosemsft commented on src/System.Net.Primitives/src/System/Net/IPAddress.cs:
```diff
@@ -609,12 +610,19 @@ public override int GetHashCode()
                 {
-                    _hashCode = StringComparer.OrdinalIgnoreCase.GetHashCode(ToString());
+                    int hashCode = HashHelpers.Combine(HashHelpers.RandomSeed, unchecked((int)PrivateScopeId));
+                    ushort[] numbers = _numbers;
```
What is the purpose of the local? Does it give better codegen?

danmosemsft replied:
Ah, I see your commit description. How do you know it produces better code -perf measurement?
@jkotas is that a known limitation, that we have to give that hint.

danmosemsft replied:
.. continuing to read backwards, I see @justinvp 's comment :) I would measure the perf.

i3arnon replied:
Indeed seems to indeed be a bit better:

|                          Method |       Mean |     Error |    StdDev |
 |-------------------------------- |-----------:|----------:|----------:|
 |             GetHashCodeToString | 153.805 ns | 2.9882 ns | 3.0687 ns |
 |          GetHashCodeHashHelpers |   7.206 ns | 0.1791 ns | 0.1917 ns |
 | GetHashCodeHashHelpersWithLocal |   5.710 ns | 0.1489 ns | 0.2087 ns |

danmosemsft replied:
Interesting. I guess it is a hint to enregister it.

danmosemsft commented:
>  '16' constant? Should we replace it with sizeof(ushort)*8?

I have seen that pattern but in our code it seems more common to have 8, 16 etc when bit shifting. I think it is fine.

stephentoub commented on src/System.Net.Primitives/src/System/Net/IPAddress.cs:
```diff
@@ -609,12 +610,19 @@ public override int GetHashCode()
+                        hashCode = HashHelpers.Combine(hashCode, numbers[i] << 16 | numbers[i + 1]);
+                    }
+
+                    _hashCode = hashCode;
```
Do we have any tests today that cover IPAddress.GetHashCode? e.g. validating that equal addresses have the same hash code, validating IPAddresses created in various ways and with various settings still successfully generate a hash code, etc.?

davidsh replied:
We absolutely need this before merge a change to GetHashCode.

morganbr commented:
This change actually _decreases_ the security of IPAddress hash codes. The string hash code algorithm is Marvin32, which is designed to make collisions difficult if you don't know the random seed. However, the HashHelpers algorithm has no such promises -- it basically just puts input bytes in order in the hash code, so it mostly just shifts the randomness off to the left. 

I do think that DoS protection for IPAddress hash codes is important since networking layers often have Dictionary<IPAddress, Something>. An attacker can forge IP addresses to cause collisions in these dictionaries. 

All that said, the string conversion is an unnecessarily slow way of getting to Marvin32. If you'd like to improve performance, you could look at pulling the managed Marvin implementation from [CoreRT](https://github.com/dotnet/corert/blob/87e58839d6629b5f90777f886a2f52d7a99c076f/src/System.Private.CoreLib/src/System/Marvin.cs) and using it here. Testing in #14354 suggests it'll be nearly as fast as HashHelpers and keep the security it has today.

danmosemsft commented:
@morganbr if https://github.com/dotnet/corefx/issues/14354 ends up with Marvin we could use that right?
 But it seems it may not as it's not the fastest and the api review explicitly stated it did not have to be DoS-Safe.

morganbr commented:
@danmosemsft , you're right on all counts. Fastest won't be the only criteria, but we haven't yet determined whether the winner will be DoS-safe. Additionally, we've reserved the right to change that algorithm. For particularly sensitive types, such as IPAddress and String, we should ensure we specifically use Marvin.

danmosemsft commented:
@morganbr should we propose another API close to the one in #14354 that is specifically Dos safe?

morganbr commented:
@danmosemsft I've been hoping to get to something like that. More specifically, I'm planning to propose a non-cryptographic hash API surface so that something that needs Marvin can just use it rather than having to copy our code around.

i3arnon commented:
I've compared the different options using BenchmarkDotNet:

 |                 Method |       Mean |     Error |    StdDev |     Median |
 |----------------------- |-----------:|----------:|----------:|-----------:|
 |    GetHashCodeToString | 164.961 ns | 3.3451 ns | 8.8124 ns | 163.162 ns |
 | GetHashCodeHashHelpers |   7.129 ns | 0.2464 ns | 0.6868 ns |   6.938 ns |
 |      GetHashCodeMarvin |  15.989 ns | 0.4259 ns | 1.2355 ns |  15.771 ns |

`Marvin` is ~2X slower than `HashHelpers` but still ~10X faster than the current implementation (and it doesn't needlessly allocates).
So I've changed the implementation to use `Marvin` to protect from DoS. As @danmosemsft suggested I used it for both IPv6 and IPv4.

I don't know of a way to code share between CoreRT and CoreFX (@morganbr, is there one?) so I've just copied over the implementation and adjusted by using `RandomNumberGenerator` instead of `Interop.GetRandomBytes`.

@stephentoub, there were some basic `GetHashCode` tests already, but I added more testing `GetHashCode` on various `IPAddress` values created in various ways.

morganbr commented on src/System.Net.Primitives/src/System/Net/IPAddress.cs:
```diff
@@ -608,30 +607,29 @@ public override bool Equals(object comparand)
             {
-                // For IPv4 addresses, we can simply use the integer representation.
-                return unchecked((int)PrivateAddress);
+                hashCode ^= Marvin.ComputeHash32(
```
Is it possible to feed the scope and address into Marvin together? I'm concerned the XOR here could be attackable. For example, if it's possible to build an IPv6 with the same scope and address bytes, they'll cancel out and hash to zero.

i3arnon replied:
I don't see how. The API accepts a `ref byte` and a count.
But I guess we can instead use the scope's hash result as the seed of the address `ComputeHash32`. That way you include both values without a XOR.

morganbr replied:
The random seeding is important (otherwise collisions become feasible). You could do something like copy to a span on the stack or maybe wrap a span around the class instance (though that makes assumptions about its layout).

morganbr replied:
On closer inspection, wrapping a span around the class doesn't work (I missed that _numbers is an array field), so that probably leaves copying. If that really doesn't work, it might be ok to XOR the scope and Marvin.DefaultSeed though it's a little hard to reason about what that does.

i3arnon replied:
I used a `Span` over a stackalloc-ed byte array and copied the Address bytes and the ScopeId onto it before calling `Marvin`.

morganbr commented on src/System.Net.Primitives/src/System/Net/IPAddress.cs:
```diff
@@ -608,30 +607,29 @@ public override bool Equals(object comparand)
-            // representation as the hashcode. Instead, we calculate
-            // the hashcode from the address literals and ScopeId.
-            if (IsIPv6)
+            if (_hashCode != 0)
```
0 is a valid value. I think that's likely ok since it would just trigger recomputation.

danmosemsft commented on src/Common/src/System/Marvin.cs:
```diff
@@ -0,0 +1,141 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
It looks like CoreRT was not testing this directly, only through String.GetHashCode.

Could you please add a few tests? They would go in corefx\src\Common\tests\System and corefx\src\Common\tests\Common.TEsts.csproj. You run them with the usual `msbuild /t:buildandtest` in the csproj folder.

Perhaps there are some canonical marvin inputs/outputs somewhere.

danmosemsft replied:
@AtsushiKan did you preserve any of the test data from when you ported this into corert?

AtsushiKan replied:
Test data is only useful if you can force the seed to a value of your own choosing. If it helps, though, here's the data I used: https://github.com/AtsushiKan/corert/tree/marvin-wip/src/Marvin

danmosemsft replied:
That's an awesome body of test cases. I would settle for 30-50 adding into corefx randomly selected cases, it should be pretty hard to break the code and still have all of them pass.

i3arnon replied:
@danmosemsft Do you mean to use these specific test data values? That would be again going indirectly through `String.GetHashCode` as these are string inputs. I can however just generate a few `byte[]` and calculate match their `Marvin` 32/64 bit hashes, but I would be doing that from the current implementation (which only tests the existing implementation). What do you prefer?

AtsushiKan replied:
@i3arnon - The test data are byte arrays - I just used hex strings to represent them in source (for succinctness - no other reason.) A `HexStringToByteArray()` utility method (not shown - but we've probably got several variants in corefx already under the `Common\Tests` subdirectory) converts them to byte arrays at runtime.

danmosemsft replied:
@i3arnon I may be misunderstanding, but it seems that @AtsushiKan was ingesting these as byte[] https://github.com/AtsushiKan/corert/blob/marvin-wip/src/Marvin/MarvinTest.cs#L18 and then hashing and comparing to the expected long result. I don't think there is any string hashing there?

The purpose of doing this would be some basic protection against someone breaking the implementation you're pasting in. 

The whole body of tests would be too big for this repo at 5MB, that would need to go into our test-data repo, I don't think you need to go to that trouble. We may expose Marvin as a public API, that would be a good time to do that.

i3arnon replied:
You're right, I missed it. I went ahead and added a test with some of this test data.

danmosemsft commented on src/System.Net.Primitives/tests/FunctionalTests/IPAddressTest.cs:
```diff
@@ -242,23 +247,55 @@ public static void Equals_Compare_Success()
+        [MemberData(nameof(GetValidIPAddresses))]
+        public static void GetHashCode_ValidIPAddresses_Success(IPAddress ipAddress)
+        {
+            ipAddress.GetHashCode();
```
I suppose there is no value in also calling this twice and checking it's equals. I see you do that above for a couple values.

stephentoub replied:
I think there's value in that.

danmosemsft commented on src/System.Net.Primitives/tests/FunctionalTests/IPAddressTest.cs:
```diff
@@ -242,23 +247,55 @@ public static void Equals_Compare_Success()
+            //192.168.0.10
+            long newIp4Address = 192 << 24 | 168 << 16 | 0 << 8 | 10;
+            ip1.Address = newIp4Address;
+            Assert.Equal("10.0.168.192", ip1.ToString());
```
is there a test anywhere that setting .Address like this also resets the hash code?

of course you can't check the hash code changes, but you can check that the new hash code is the same as a fresh IPAddress created over the new address

danmosemsft replied:
I think this bug already existed for ipv6, and now it is for ipv4 as well. Setting Address should clear _hashCode.

i3arnon replied:
I don't think this bug existed for ipv6 because you can't set `Address` for ipv6. However, i do think there's a similar bug with `ScopeId`. You can set it for ipv6 and it's part of the string representation (hence, also part of the current hash code) but the cached `_toString` doesn't get reset when `ScopeId` is set.
I now reset both `_toString` and `_hashCode` in the private address and scope id setters. I hope that's ok.

stephentoub commented on src/System.Net.Primitives/tests/FunctionalTests/IPAddressTest.cs:
```diff
@@ -242,23 +247,55 @@ public static void Equals_Compare_Success()
+        [MemberData(nameof(GeneratedIPAddresses))]
+        public static void GetHashCode_VariousIPAddressGenerators_Success(IPAddress ipAddress)
+        {
+            ipAddress.GetHashCode();
```
Ditto; I think it makes sense as Dan suggested to change this to something like:
```C#
Assert.Equal(ipAddress.GetHashCode(), ipAddress.GetHashCode());
```

AtsushiKan commented:
I'd recommend sampling the test data via these principles:

- Make sure the lengths of the input bytes are distributed such that each of the 8 "case" blocks in the switch statement inside Marvin are exercised.

- Make sure there's representation for "0x80 bit" on and "0x80 bit" off in each of the position in the input bytes. This is to catch "oops - sign-extended when I should have zero-extended" bugs.

i3arnon commented on src/Common/tests/Tests/System/MarvinTests.cs:
```diff
@@ -0,0 +1,68 @@
+            Assert.Equal((long)expectedHash, hash);
+        }
+
+        public static object[][] TestDataAndExpectedHashes =
```
@AtsushiKan Is this what you meant?

AtsushiKan replied:
Yeah, that's what I was thinking.

danmosemsft commented on src/Common/tests/Tests/System/MarvinTests.cs:
```diff
@@ -0,0 +1,68 @@
+
+        public static object[][] TestDataAndExpectedHashes =
+        {
+            new object[] { 0x4FB61A001BDBCCLU, TestDataString0Byte, 0x30ED35C100CD3C7DLU },
```
Nit, it might be tidier to extract constants seed1, seed2 ,seed3.

danmosemsft commented:
BTW, seems like you're squashing and force pushing or equivalent - that's OK but it's usually good to avoid as it means we can't review just the changes since your last commit. :)

stephentoub commented on src/System.Net.Primitives/src/System/Net/IPAddress.cs:
```diff
@@ -612,23 +613,45 @@ public override bool Equals(object comparand)
                 {
-                    _hashCode = StringComparer.OrdinalIgnoreCase.GetHashCode(ToString());
+                    var addressSpan = new ReadOnlySpan<byte>(Unsafe.AsPointer(ref _numbers[0]), IPAddressParserStatics.IPv6AddressBytes);
+                    var scopeIdSpan = new ReadOnlySpan<byte>(Unsafe.AsPointer(ref _addressOrScopeId), sizeof(uint));
```
@jkotas, is this safe?  I'd have thought the GC wouldn't scan the ptr resulting from Unsafe.AsPointer, but does it?

jkotas replied:
This is not safe. This code has GC hole.

i3arnon replied:
I've used `new ReadOnlySpan<ushort>(_numbers).AsBytes()` instead. I assume that's safe as it doesn't go through a pointer, is it?

stephentoub replied:
Yes, that's fine.

i3arnon commented:
@danmosemsft indeed it's a deeper change than I would have guessed... but that's a good thing :)

stephentoub commented on src/System.Net.Primitives/src/System/Net/IPAddress.cs:
```diff
@@ -612,23 +615,44 @@ public override bool Equals(object comparand)
                 {
-                    _hashCode = StringComparer.OrdinalIgnoreCase.GetHashCode(ToString());
+                    byte* addressAndScopeId = stackalloc byte[addressAndScopeIdLength];
+                    addressAndScopeIdSpan = new Span<byte>(addressAndScopeId, addressAndScopeIdLength);
```
corefx is using a preview 7.2 compiler that has additional span-related features.  This can be changed to just:
```C#
Span<byte> addressAndScopeIdSpan = stackalloc byte[addressAndScopeIdLength];
```

stephentoub commented on src/System.Net.Primitives/src/System/Net/IPAddress.cs:
```diff
@@ -612,23 +615,44 @@ public override bool Equals(object comparand)
-                return _hashCode;
+                new ReadOnlySpan<ushort>(_numbers).AsBytes().CopyTo(addressAndScopeIdSpan);
+                Span<byte> scopeIdSpan = addressAndScopeIdSpan.Slice(IPAddressParserStatics.IPv6AddressBytes);
+                BitConverter.TryWriteBytes(scopeIdSpan, _addressOrScopeId);
```
Nit: Consider adding an assert to ensure this returns true, e.g.
```C#
bool scopeWritten = BitConverter.TryWriteBytes(scopeIdSpan, _addressOrScopeId);
Debug.Assert(scopeWritten);
```

stephentoub commented on src/Common/src/System/Marvin.cs:
```diff
@@ -0,0 +1,141 @@
+
+        public static ulong DefaultSeed => s_defaultSeed;
+
+        private static ulong s_defaultSeed = GenerateSeed();
```
Nit: readonly

stephentoub replied:
Or:
```C#
public static ulong DefaultSeed { get; } = GenerateSeed();
```

stephentoub commented on src/Common/src/System/Marvin.cs:
```diff
@@ -0,0 +1,141 @@
+            {
+                var bytes = new byte[sizeof(ulong)];
+                rng.GetBytes(bytes);
+                return BitConverter.ToUInt64(bytes, 0);
```
Nit: It's not a big deal as this is done only once, but this could be done without allocating the byte[], now that both RandomNumberGenerator and BitConverter have span support, e.g.
```C#
Span<byte> bytes = stackalloc byte[ulong];
rng.GetBytes(bytes);
return BitConverter.ToUInt64(bytes);
```

stephentoub commented:
Hmm, the failures appear to be because or common test project is targeting netstandard rather than netcoreapp. We should probably just change that to netcoreapp... I don't think there's a benefit in having it be netstandard.

i3arnon commented:
@stephentoub The build still fails, I assume because `Marvin` is also compiled in `System.Net.Primitives` and the tests target netstandard. Should I change these as well or go back to `byte[]`?

stephentoub commented:
> The build still fails

Ugh, all these configurations make my head spin.  Since it's a one-time allocation, go ahead and back it out for now, but I think we should revisit it subsequently.

stephentoub commented:
@dotnet-bot test UWP CoreCLR x64 Debug Build please (Task test failure, unrelated)

stephentoub commented:
Thanks for proposing and implementing this, @i3arnon.

i3arnon commented:
Would it be possible to consider this change for porting to the full FX?

karelz commented:
Anything is possible :). It is just matter of cost vs. benefit (vs. risk).
What is the benefit for real-world end-to-end scenarios? You mentioned it showed up high in your app - how high? Is your app unique? How many other apps/libraries do we expect in that boat?

i3arnon commented:
I expect this to only affect very data-intensive network-related applications (i.e. many different IPAddress instances). Those are pretty rare (especially written in managed code).

For us though ([Microsoft ATA](https://microsoft.com/ata)) it's pretty high up as we parse a constant stream of network traffic and we've already optimized a lot of the low hanging fruits.

It's far from critical for us... but it would be nice-to-have.

karelz commented:
Can you estimate the savings in % of your compute time (or in $$)?
Making large .NET services efficient and successful is a benefit we could consider worthwhile.

i3arnon commented:
I recently finished a stretch of allocation reduction so I'll need to run some tests to see where IPAddress ends up on the list.
It's usually somewhere in the top 10 allocating stacks for us, but I'm not sure about percentages.
